File: c/totientRange.c
// TotientRange.c - Sequential Euler Totient Function (C Version)
// compile: gcc -Wall -O -o TotientRange TotientRange.c
// run:     ./TotientRange lower_num upper_num

// Greg Michaelson 14/10/2003
// Patrick Maier   29/01/2010 [enforced ANSI C compliance]

// This program calculates the sum of the totients between a lower and an 
// upper limit using C longs. It is based on earlier work by:
// Phil Trinder, Nathan Charles, Hans-Wolfgang Loidl and Colin Runciman

// The comments provide (executable) Haskell specifications of the functions

#include <stdio.h>
#include <omp.h>
#include <math.h>
#include <sys/time.h>

// hcf x 0 = x
// hcf x y = hcf y (rem x y)

long hcf(long x, long y)
{
  long t;

  while (y != 0) {
    t = x % y;
    x = y;
    y = t;
  }
  return x;
}


// relprime x y = hcf x y == 1

int relprime(long x, long y)
{
  return hcf(x, y) == 1;
}


// euler n = length (filter (relprime n) [1 .. n-1])

long euler(long n)
{
  long length, i;

  length = 0;
  for (i = 1; i < n; i++)
    if (relprime(n, i))
      length++;
  return length;
}


// sumTotient lower upper = sum (map euler [lower, lower+1 .. upper])

long sumTotient(long lower, long upper, int num_threads)
{
  long sum, i;
  long eulers[(upper - (lower-1))];

  omp_set_num_threads(num_threads);

  #pragma omp parallel for \
  shared(eulers) private(i) \
  schedule(dynamic)
  for (i = lower; i <= upper; i++) {
    eulers[i-1] = euler(i);
  }
    

  sum = 0;
  long j;
  for (j = 0; j < upper; j++ ) {
      sum += eulers[j];
   }

  return sum;
}


int main(int argc, char ** argv)
{
  long lower, upper;
  int num_threads;

  if (argc != 4) {
    printf("not 3 arguments\n");
    return 1;
  }
  sscanf(argv[1], "%ld", &lower);
  sscanf(argv[2], "%ld", &upper);
  sscanf(argv[3], "%d", &num_threads);


  long totientSum;
  struct timeval start, end;

  gettimeofday(&start, NULL);
  totientSum = sumTotient(lower, upper, num_threads);
  gettimeofday(&end, NULL);

  double time_taken = end.tv_sec + end.tv_usec / 1e6 -
                        start.tv_sec - start.tv_usec / 1e6; // in seconds

  printf("C: Sum of Totients  between [%ld..%ld] is %ld\n", lower, upper, totientSum);
  printf("Elapsed time %fs\n", time_taken);
  return 0;
}

File: go/totientRange/totientRange.go
// totientRange.go - Parallel Euler Totient Function (Go Version)
// compile: go build
// run:     totientRange lower_num upper_num

// Chase Condon	   10/11/2019
// Modified from sequential version by Phil Trinder

// This program calculates the sum of the totients between a lower and an
// upper limit
//
// Each function has an executable Haskell specification
//
// It is based on earlier work by: Greg Michaelson, Patrick Maier, Phil Trinder,
// Nathan Charles, Hans-Wolfgang Loidl and Colin Runciman

package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
	"sync"
)

// Compute the Highest Common Factor, hcf of two numbers x and y
//
// hcf x 0 = x
// hcf x y = hcf y (rem x y)

func hcf(x, y int64) int64 {
	var t int64
	for y != 0 {
		t = x % y
		x = y
		y = t
	}
	return x
}

// relprime determines whether two numbers x and y are relatively prime
//
// relprime x y = hcf x y == 1

func relprime(x, y int64) bool {
	return hcf(x, y) == 1
}

// euler(n) computes the Euler totient function, i.e. counts the number of
// positive int64egers up to n that are relatively prime to n
//
// euler n = length (filter (relprime n) [1 .. n-1])

func euler(n int64) int64 {
	var length, i int64

	length = 0
	for i = 1; i < n; i++ {
		if relprime(n, i) {
			length++
		}
	}
	return length
}

// helper function for workers 
// pulls jobs from the job pool and runs the Euler function 
func worker(id int, jobs <-chan int64, results chan<- int64, wg *sync.WaitGroup) {
	for j := range jobs {
		results <- euler(j)
		wg.Done()
	}
}

// sumTotient lower upper sums the Euler totient values for all numbers
// between "lower" and "upper".
// 
// creates a pool of worker threads and gives them a job channel
// which is populated with numbers in the range [lower, upper]
//
// sumTotient lower upper = sum (map euler [lower, lower+1 .. upper])

func sumTotient(lower, upper, num_threads int64) int64 {
	var wg sync.WaitGroup
	jobs := make(chan int64, upper)
	results := make(chan int64, upper)
	var sum, i int64

	for w := 1; w <= int(num_threads); w++ {
		go worker(w, jobs, results, &wg)
	}

	for i = lower; i <= upper; i++ {
		wg.Add(1)
		jobs <- i
	}
	close(jobs)
	wg.Wait()
	close(results)

	sum = 0
	for result := range results {
		sum += result
	}
	
	return sum
}

func main() {
	var lower, upper, num_threads int64
	var err error

	// Read and validate lower and upper arguments
	if len(os.Args) < 4 {
		panic(fmt.Sprintf("Usage: must provide lower, upper range limits, and number of threads as arguments"))
	}

	if lower, err = strconv.ParseInt(os.Args[1], 10, 64); err != nil {
		panic(fmt.Sprintf("Can't parse first argument"))
	}
	if upper, err = strconv.ParseInt(os.Args[2], 10, 64); err != nil {
		panic(fmt.Sprintf("Can't parse second argument"))
	}
	if num_threads, err = strconv.ParseInt(os.Args[3], 10, 64); err != nil {
		panic(fmt.Sprintf("Can't parse third argument"))
	}

	start := time.Now()

	totientSum := sumTotient(lower, upper, num_threads)

	fmt.Println("Sum of Totients between", lower, "and", upper, "is", totientSum)
	elapsed := time.Since(start)
	fmt.Println("Elapsed time", elapsed.Seconds())
}
